lib/jsdom/living/helpers/http-request.js

```diff
@@ -0,0 +1,272 @@
+            const buffer = buffers[i++];
+            if (!message.writableEnded) {
+              message.write(buffer.data, buffer.encoding, writeNext);
+            }
+          } else if (self._ended) {
+            message.end();
+          }
+        }
+      })();
+    }
+  }
+
+  _processResponse(response) {
+    response.request = {};
+    response.request.headers = this._options.headers;
+    response.request.uri = new URL(this._currentUrl);
+    this.response = response;
+    this.setCookies(response);
+    const { statusCode } = response;
+    const { location } = response.headers;
+    const catchResErrors = err => {
+      if (!(majorNodeVersion === 15 && err.message === "aborted")) {
+        this.emit("error", err);
+      }
+    };
+    response.on("error", catchResErrors);
+    let redirectAddress = null;
+    if (typeof location === "string" &&
+      location.length &&
+      this._options.followRedirects &&
+      statusCode >= 300 &&
+      statusCode < 400) {
+      redirectAddress = location;
+    } else if (statusCode === 401 &&
+      /^Basic /i.test(response.headers["www-authenticate"] || "") &&
+      (this._options.user && this._options.user.length)) {
+      const { user, pass } = this._options;
+      this._options.auth = `${user}:${pass}`;
+      redirectAddress = 1;
+    }
+    if (redirectAddress) {
+      if (++this._redirectCount > 21) {
+        const redirectError = new Error("Maximum number of redirects exceeded");
+        redirectError.code = "ERR_TOO_MANY_REDIRECTS";
+        this.emit("error", redirectError);
+        return;
+      }
+      abortRequest(this._currentRequest);
+      response.destroy();
+      this._isRedirect = true;
+      if (((statusCode === 301 || statusCode === 302) && this._options.method === "POST") ||
+        (statusCode === 303 && !/^(?:GET|HEAD)$/.test(this._options.method))) {
+        this._options.method = "GET";
+        this._requestBodyBuffers = [];
+      }
+      let previousHostName = this.removeMatchingHeaders(/^host$/i);
+      if (!previousHostName) {
+        previousHostName = new URL(this._currentUrl).hostname;
+      }
+      const previousURI = this._currentUrl;
+      if (redirectAddress !== 1) {
+        const nextURL = redirectAddress.startsWith("https:") ?
+          new URL(redirectAddress) :
+          new URL(redirectAddress, this._currentUrl);
+        if (nextURL.hostname !== previousHostName) {
+          this.removeMatchingHeaders(/^authorization$/i);
+        }
+        this._currentUrl = nextURL.toString();
+      }
+      this._options.headers.Referer = previousURI;
+      this.response.request.uri = new URL(this._currentUrl);
+      this.emit("redirect");
+      try {
+        this._performRequest();
+      } catch (cause) {
+        this.emit("error", cause);
+      }
+    } else {
+      const isCompressed = this._options.headers["Accept-Encoding"] === "gzip, deflate";
+      const self = this;
+      function hasBody(code) {
```

Comment:
This function is only called once and does the ugly `self = this` thing. It's best to just inline.